diff -ruw dracut.git-orig/modules.d/99base/init dracut.git/modules.d/99base/init
--- dracut.git-orig/modules.d/99base/init	2012-02-13 16:47:10.994670000 +0000
+++ dracut.git/modules.d/99base/init	2012-02-13 16:47:24.361802000 +0000
@@ -8,8 +8,6 @@
 # Harald Hoyer <harald@redhat.com>
 # Jeremy Katz <katzj@redhat.com>
 
-export -p > /tmp/export.orig
-
 wait_for_loginit()
 {
     set +x
@@ -366,8 +364,6 @@
             unset "$i";;
     esac
 done
-. /tmp/export.orig 2>/dev/null || :
-rm -f /tmp/export.orig
 
 initargs=""
 read CLINE </proc/cmdline
